src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
>I don't understand why constructor selection needs to be well defined.

Because without a uniform expected method of grain constructor selection, we've no means of having any confidence that the native Orleans grains nor grains in extension libraries will be created correctly with user provided containers.

>ASP.NET in particular does not put this constraint

Yes it does.  Conforming containers, which is all we support, have well defined constructor selection, which is what we should follow.
See DependencyInjectionSpecificationTests.ServiceContainerPicksConstructorWithLongestMatches
Current facet code does not conform to this as it requires a single public constructor, but we can address that as the system matures.  This is just the initial implementation.

>What is the need to couple it to GrainTypeData

Just a convenient place to cache grain type specific data.  This is an internal implementation detail.  If we find a better place we can move it.